walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,235 @@
+    Collections.sort(allTimeRanges, TimeRange.ORDER_BY_START); 
+  }
+
+  /**
+   * Finds all free time ranges based on a list of busy time ranges and 
+   * the length of the requested meeting
+   */ 
+  private List<TimeRange> getFreeTimeRanges(List<TimeRange> allBusyTimeRanges, long meetingDurationMinutes) {
+    List<TimeRange> freeTimeRanges = new ArrayList<TimeRange>();
+
+    if (allBusyTimeRanges.isEmpty() && meetingDurationMinutes <= TimeRange.END_OF_DAY) {
+      freeTimeRanges.add(TimeRange.fromStartEnd(
+          /* start= */ 0, 
+          /* end= */ TimeRange.END_OF_DAY, 
+          /* inclusive= */ true));
+      return freeTimeRanges; 
+    }
+
+    // A for loop is necessary because each consecutive pair 
+    // of time ranges needs to be looked at at each iteration. 
+    for (int i = 0; i < allBusyTimeRanges.size(); i++) {
+      if (i == 0) {
+        int earliestStart = allBusyTimeRanges.get(i).start();
+        if (earliestStart >= meetingDurationMinutes) {
+          freeTimeRanges.add(TimeRange.fromStartEnd(
+            /* start= */ TimeRange.START_OF_DAY, 
+            /* end= */ earliestStart, 
+            /* inclusive= */ false)); 
+        }
+      } 
+
+      if (i == allBusyTimeRanges.size() - 1) {
+        int latestEnd = allBusyTimeRanges.get(i).end();
+        if ((TimeRange.END_OF_DAY - latestEnd) >= meetingDurationMinutes) {
+          // According to TimeRange class's implementation of the time ranges,
+          // TimeRange.END_OF_DAY variable should have "true" for the 
+          // "inclusive" parameter. 
+          freeTimeRanges.add(TimeRange.fromStartEnd(
+            /* start= */ latestEnd, 
+            /* end= */ TimeRange.END_OF_DAY, 
+            /* inclusive= */ true)); 
+        }
+      } else {
+        int currEnd = allBusyTimeRanges.get(i).end();
+        int nextStart = allBusyTimeRanges.get(i + 1).start(); 
+
+        if (nextStart - currEnd >= meetingDurationMinutes) {
+          freeTimeRanges.add(TimeRange.fromStartEnd(
+            /* start= */ currEnd, 
+            /* end= */ nextStart, 
+            /* inclusive= */ false)); 
+        }
+      }
+    } 
+    return freeTimeRanges; 
+  }
+
+  /** Finds the common time range between time ranges of two collections */ 
+  private List<TimeRange> findCommonFreeTimeRanges(
+      List<TimeRange> mandatoryAttendeesFreeTimeRanges, 
+      List<TimeRange> optionalAttendeesFreeTimeRanges, 
+      long meetingDurationMinutes) {
+        
+      // If any of mandatory or optional attendee has no free time, 
+      // return the other group's free time slots. 
+      if (mandatoryAttendeesFreeTimeRanges.isEmpty()) {
+        return optionalAttendeesFreeTimeRanges;
+      }
+      if (optionalAttendeesFreeTimeRanges.isEmpty()) {
+        return mandatoryAttendeesFreeTimeRanges; 
+      }    
+
+      List<TimeRange> commonTimeRanges = new ArrayList<TimeRange>();
+      int i = 0;
```

Comment:
Consider adding some comment that explains that i is for mandatory attendees and j is for optional attendees.

Reply:
Updated.